drivers/sensor/sgp40/sgp40.c

```diff
@@ -0,0 +1,309 @@
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
+{
+	return -ENOTSUP;
+}
+
+static int sgp40_sample_fetch(const struct device *dev,
+			       enum sensor_channel chan)
+{
+	struct sgp40_data *data = dev->data;
+	const struct sgp40_config *cfg = dev->config;
+
+	uint8_t rx_buf[3];
+	uint16_t raw_sample;
+
+	if (chan != SENSOR_CHAN_GAS_RES && chan != SENSOR_CHAN_ALL) {
+		return -ENOTSUP;
+	}
+
+	if (sgp40_start_measurement(dev)) {
+		LOG_ERR("Failed to start measurement.");
+		return -EIO;
+	}
+
+	k_sleep(K_MSEC(SGP40_MEASURE_WAIT_MS));
+
+	if (i2c_read(cfg->bus, rx_buf, sizeof(rx_buf), cfg->i2c_addr)) {
+		LOG_ERR("Failed to read data sample.");
+		return -EIO;
+	}
+
+	raw_sample = sys_get_be16(rx_buf);
+ 	if (sgp40_compute_crc(raw_sample) != rx_buf[2]) {
+		LOG_ERR("Invalid CRC for data sample.");
+		return -EIO;
+	}
+
+	data->raw_sample = raw_sample;
+
+	return 0;
+}
+
+static int sgp40_channel_get(const struct device *dev,
+			      enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	const struct sgp40_data *data = dev->data;
+
+	if (chan != SENSOR_CHAN_GAS_RES) {
+		return -ENOTSUP;
+	}
+
+	val->val1 = data->raw_sample;
+	val->val2 = 0;
+
+	return 0;
+}
+
+
+#ifdef CONFIG_PM_DEVICE
+static int sgp40_set_power_state(const struct device *dev,
+				  uint32_t power_state)
+{
+	struct sgp40_data *data = dev->data;
+	uint16_t cmd;
+
+	if (data->pm_state == power_state) {
+		LOG_ERR("Device already in requested PM_STATE.");
+		return 0;
+	}
+
+	if (power_state == PM_DEVICE_STATE_ACTIVE) {
+		/* activate the hotplate by sending a measure command */
+		cmd = SGP40_CMD_MEASURE_RAW;
+	} else if (power_state == PM_DEVICE_STATE_LOW_POWER ||
+			power_state == PM_DEVICE_STATE_SUSPEND) {
+		cmd = SGP40_CMD_HEATER_OFF;
+	} else {
+		LOG_DBG("Power state not implemented.");
+		return -ENOSYS;
+	}
+
+	if (sgp40_write_command(dev, cmd)){
+		LOG_ERR("Failed to set power state.");
+		return -EIO;
+	}
+
+	data->pm_state = power_state;
+
+	return 0;
+}
+
+static uint32_t sgp40_get_power_state(const struct device *dev,
+		uint32_t *state)
+{
+	struct sgp40_data *data = dev->data;
+
+	*state = data->pm_state;
+
+	return 0;
+}
+
+static int sgp40_pm_ctrl(const struct device *dev,
+	uint32_t ctrl_command,
+	uint32_t *state,
+	pm_device_cb cb,
+	void *arg)
+
+{
+	int rc = 0;
+
+	if (ctrl_command == PM_DEVICE_STATE_SET) {
+		rc = sgp40_set_power_state(dev, *state);
+	} else if (ctrl_command == PM_DEVICE_STATE_GET) {
+		rc = sgp40_get_power_state(dev, state);
+	}
+
+	if (cb) {
+		cb(dev, rc, state, arg);
+	}
+	return rc;
+}
+#endif /* CONFIG_PM_DEVICE */
+
+static int sgp40_init(const struct device *dev)
+{
+	const struct sgp40_config *cfg = dev->config;
+	struct sgp40_compensation_data comp_data;
+	comp_data.t = 25;
+	comp_data.rh = 50;
+
+	if (!device_is_ready(cfg->bus)) {
+		LOG_ERR("Device not ready.");
+		return -ENODEV;
+	}
+
+	if (cfg->selftest) {
+		if (sgp40_selftest(dev)) {
```

Comment:
error codes should be propagated

Reply:
i've updated this across the sht4x driver aswell